js/src/gtag-events/utils.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Internal dependencies
+ */
+import { sendToGroup } from './constants';
+
+/* global glaGtagData */
+
+/**
+ * Track an event using the global gtag function.
+ *
+ * @param {string} eventName
+ * @param {Object} eventParams
+ */
+export const trackEvent = ( eventName, eventParams ) => {
+	if ( typeof gtag !== 'function' ) {
+		throw new Error( 'Function gtag not implemented.' );
+	}
+	// eslint-disable-next-line no-console
+	console.log( `Tracking event ${ eventName }` );
+	window.gtag( 'event', eventName, {
+		send_to: sendToGroup,
+		...eventParams,
+	} );
+};
+
+/**
+ * Track an add_to_cart event.
+ *
+ * @param {Object} product
+ * @param {number} quantity
+ */
+export const trackAddToCartEvent = ( product, quantity = 1 ) => {
+	trackEvent( 'add_to_cart', {
+		ecomm_pagetype: 'cart',
+		event_category: 'ecommerce',
+		items: [ getItemObject( product, quantity ) ],
+	} );
+};
+
+/**
+ * Formats data into a cart Item object.
+ *
+ * @param {Object} product
+ * @param {number} quantity
+ * @return {Object} Item object.
+ */
+export const getItemObject = ( product, quantity ) => {
+	const item = {
+		id: 'gla_' + product.id,
+		quantity,
+		google_business_vertical: 'retail',
+	};
+
+	if ( product.name ) {
+		item.name = product.name;
+	}
+
+	if ( 'categories' in product && product.categories.length ) {
+		item.category = product.categories[ 0 ].name;
+	}
+
+	if ( product.prices && product.prices.price ) {
+		item.price =
+			parseInt( product.prices.price, 10 ) /
+			10 ** product.prices.currency_minor_unit;
+	}
+
+	return item;
+};
+
+/**
+ * Formats a regular price into a price object.
+ *
+ * @param {number} price
+ * @return {Object} Price object.
+ */
+export const getPriceObject = ( price ) => {
+	return {
+		price: price * 10 ** glaGtagData.currency_minor_unit,
+		currency_minor_unit: glaGtagData.currency_minor_unit,
+	};
+};
+
+/**
+ * Formats a product object to include name and price from global data.
+ *
+ * @param {Object} product
+ * @return {Object} Product object.
+ */
+export const getProductObject = ( product ) => {
+	if ( product.id && glaGtagData.products[ product.id ] ) {
+		product.name = glaGtagData.products[ product.id ].name;
+		product.prices = getPriceObject(
+			glaGtagData.products[ product.id ].price
+		);
+	}
+	return product;
+};
+
+/**
+ * Updates product data with the retrieved variation.
+ *
+ * @param {Object} variation
+ */
+export const retrievedVariation = ( variation ) => {
```

Comment:
💅  Since `retrievedVariation` seems to be triggered by a Jquery event somewhere, I would make sure variation is a valid object here as well.

Reply:
What kind of validation do you suggest here? Just a simple check if `variation.variation_id` is set or should we check for each parameter we use? At what kind of level can we trust the data that is sent back and forth, or it better to always assume it can be garbled?

Reply:
I guess in this case just checking `variation.variation_id` to be sure variation is there an is an object would be enough to avoid a crash in case variation is not there or is not an object.  In case display_name and display_price I won't check that... since it won't make the app to crash and in the worse case scenario we will send the props as undefined. 

`if (! variation?.variation_id ) return` before line 106 

> At what kind of level can we trust the data that is sent back and forth, or it better to always assume it can be garbled?

In this case since the data is coming from a JQuery event `found_variation` that we don't control I would add the check. But again this is a very 💅  opinionated question based on how defensive your are.